ProducerConsumer.cpp

```diff
@@ -49,25 +73,28 @@ class ProducerConsumer
 	using ProducerFunc = std::function<T()>;
 	using ConsumerFunc = std::function<void(const T&)>;
 
-	ProducerConsumer(ProducerFunc produce, ConsumerFunc consume)
-		: produce(produce),
-		consume(consume)
-	{}
+	ProducerConsumer(ProducerFunc produce, ConsumerFunc consume) : produce(produce), consume(consume) {}
 
 	void run(int itemCount)
 	{
-		while (itemCount-->0)
+#pragma omp parallel
 		{
-			T item = produce();
-			buffer.push(item);
-
-			item = buffer.pop();
-			consume(item);
+#pragma omp for
+			for (int i = 0; i < itemCount; i++)
+			{
+				T item = produce();
+#pragma omp critical
```

Comment:
buffer here is synchronized queue implemented exactly for concurrent access from different threads. Push and pop methods are intended to be used without additional synchronization, because buffer has already its own underlying synchronization mechanism. 
This omp critical looks redundant therefore. Try how it works without critical section, but just
```
buffer.push()
if (!buffer.empty)
    buffer.pop()
```

Reply:
I supposed that, but I did not sure that it works exactly like that way. I have tried your code, and it works well. (My new commit is on the git)